### PR TITLE
Introduce Table Property  `delta.dataSkippingStatsColumns`

### PR DESCRIPTION
**Pull request by kamcheungting-db**
## Description
This PR introduces new table property `dataSkippingStatsColumns ` for users to specify the set of columns that collect file skipping statistics.

The syntax of setting this table property is:
```
CREATE TABLE <Table Name> (<Column Definition>)
TBLPROPERTIES('delta.dataSkippingStatsColumns' = '[<column-identifier-1>, …, <column-identifier-N>]');

ALTER TABLE <Table Name>
SET TBLPROPERTIES ('delta.dataSkippingStatsColumns' = '[<column-identifier-1>, …, <column-identifier-N>]');
```

The CREATE TABLE and Alter command handler would validates that all `delta.dataSkippingStatsColumns` exists.
These two commands would error out, when:
- If any column of `delta.dataSkippingStatsColumns` doesn't exist,
- If any column of `delta.dataSkippingStatsColumns` doesn't support data skipping,
- If any column of `delta.dataSkippingStatsColumns` is partitioned column.

If user drops a column, the corresponding entry inside `delta.dataSkippingStatsColumns` would also be removed.

If user renames a column, the corresponding entry inside `delta.dataSkippingStatsColumns` would also be renamed.

The OPTIMIZE ZORDER command also recognizes the `delta.dataSkippingStatsColumns`.

The `delta.dataSkippingStatsColumns` contains a list of column identifiers. Each column identifier is represented as:
## column-identifier
An identifier is a string used to identify a database object such as a table, view, schema, column. Both regular identifiers and delimited identifiers are case-insensitive. The Regular Column Identifier contains following characters: `{ letter | digit | '_' } [ , ... ]`. If there is special characters, which include `!@#$%^&*()_+-={}|[]\:";'<>,.?/`, inside the column identifier ` should be used to escape the column name.

#### CREATE TABLE
```
CREATE TABLE T1 (c0 long, `c-1` long, `c!@#$`) using delta
TBLPROPERTIES('delta.dataSkippingStatsColumns' = 'c0, `c-1`, `c!@#$`');
```
#### ALTER TABLE
```
CREATE TABLE T2 (c0 long, `c-1` long, `c!@#$`) using delta;

ALTER TABLE T2 SET TBLPROPERTIES ('delta.dataSkippingStatsColumns' = 'c0, `c-1`, `c!@#$`');
```
#### NESTED COLUMN
```
CREATE TABLE T3 (c0 long, c1 STRUCT <c11: String, c12: long, c13 STRUCT <c131: long, c132: long>>) using delta
TBLPROPERTIES('delta.dataSkippingStatsColumns' = 'c0, c1.c11, c1.c12, c1.c13.c131, c1.c13.c132');

CREATE TABLE T4 (c0 long, c1 STRUCT <c11: String, c12: long, c13 STRUCT <c131: long, c132: long>>) using delta
TBLPROPERTIES('delta.dataSkippingStatsColumns' = 'c0, c1.c11, c1.c12, c1.c13');

CREATE TABLE T5 (c0 long, c1 STRUCT <c11: String, c12: long, c13 STRUCT <c131: long, c132: long>>) using delta
TBLPROPERTIES('delta.dataSkippingStatsColumns' = 'c0, c1');
```

## How was this patch tested?
Unit Test:
1. validates setting table property procedure:
  1.1. detects non-existing columns,
  1.2. support nested columns,
  1.3. detects unsupported column types,
  1.4. handles columns with special characters,
  1.5. handles columns with invalid datatype in both nested and un-nested column,
  1.6. handles partition columns in both nested and un-nested column.
  1.7. handles duplicated columns in both nested and un-nested columns.
2. handle drop columns
2.1. drop nested columns from a table
2.2. drop flat columns from a table
3. handle rename columns
3.1. rename nested columns from a table
3.2. rename flat columns from a table
4. optimize z-order can recognize delta statistics columns.
5. All column mapping modes are supported.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala [hunk 1]
```diff
@@ -389,7 +384,306 @@ trait StatisticsCollection extends DeltaLogging {
...
+      statsColPaths: Seq[Seq[String]],
+      mappingMode: DeltaColumnMappingMode,
+      parentPath: Seq[String] = Seq.empty): Option[StructType] = {
+    // Find the unique column names at this nesting depth, each with its path remainders (if any)
+    val cols = statsColPaths.groupBy(_.head).mapValues(_.map(_.tail))
+    val newSchema = schema.flatMap { field =>
+      cols.get(field.name).flatMap { paths =>
+        field.dataType match {
+          case _ if paths.forall(_.isEmpty) =>
+            val fullPath = (parentPath:+ field.name).mkString(".")
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:607]: does this kind of concatenating ... handle field name with . in them?

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala [hunk 2]
```diff
@@ -389,7 +384,306 @@ trait StatisticsCollection extends DeltaLogging {
...
+  def filterSchema(
+      schemaNames: Seq[String],
+      schema: StructType,
+      statsColPaths: Seq[Seq[String]],
+      mappingMode: DeltaColumnMappingMode,
+      parentPath: Seq[String] = Seq.empty): Option[StructType] = {
+    // Find the unique column names at this nesting depth, each with its path remainders (if any)
+    val cols = statsColPaths.groupBy(_.head).mapValues(_.map(_.tail))
+    val newSchema = schema.flatMap { field =>
+      cols.get(field.name).flatMap { paths =>
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:604]: so if the field name in the property is "x.y.z" and `x` is the `field` here.. then `paths` is `[y, z]` right?

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala [hunk 3]
```diff
@@ -389,7 +384,306 @@ trait StatisticsCollection extends DeltaLogging {
...
+        field.dataType match {
+          case _ if paths.forall(_.isEmpty) =>
+            val fullPath = (parentPath:+ field.name).mkString(".")
+            def convertToPhysicalName(field: StructField): StructField = {
+              // If mapping mode is NoMapping or the dataSchemaName already contains the mapped
+              // column name, the schema mapping can be skipped.
+              if (mappingMode == NoMapping || schemaNames.contains(fullPath)) return field
+              val physicalName = field.metadata.getString("delta.columnMapping.physicalName")
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:612]: this constant should be defined somewhere. its likely this is being used multiple places.

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala [hunk 4]
```diff
@@ -389,7 +384,306 @@ trait StatisticsCollection extends DeltaLogging {
...
+      }
+  }
+
+  /**
+   * Generates a truncated data schema for stats collection.
+   * @param schema the original data schema
+   * @param indexedCols the maximum number of leaf columns to collect stats on
+   * @return truncated schema and the number of leaf columns in this schema
+   */
+  def truncateSchema(schema: StructType, indexedCols: Int): (StructType, Int) = {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:666]: this was private before.. does it need to be public?

**Review by tdas — CHANGES_REQUESTED** [review]
overall this looks great! this really adds a new flexibility for delta users to customize their stats collection.

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala [hunk 5]
```diff
@@ -490,6 +490,23 @@ trait DeltaConfigsBase extends DeltaLogging {
     a => a >= -1,
     "needs to be larger than or equal to -1.")
 
+  /**
+   * The names of specific columns to collect stats on for data skipping. If present, it takes
+   * precedences over dataSkippingNumIndexedCols config, and the system will only collect stats for
```
[the remark this answers is not in the document]
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala:495]: Done

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala (hunk 4 above)
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:666]: Made it private.

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala (hunk 3 above)
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:612]: Done

**Comment by Tagar** [comment]
Would Delta collect stats on the first 32 columns and columns specified in deltaStatsColumns?

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala [hunk 6]
```diff
@@ -490,6 +490,23 @@ trait DeltaConfigsBase extends DeltaLogging {
...
+   * stats schema of table, i.e., it will change the behavior of future stats collection (e.g., in
+   * append and OPTIMIZE) as well as data skipping (e.g., the column stats not mentioned by this
+   * config will be ignored even if they exist).
+   */
+  val DATA_SKIPPING_STATS_COLUMNS = buildConfig[Option[String]](
+    "dataSkippingStatsColumns",
+    null,
+    v => Option(v),
+    vOpt => vOpt.forall(v => StatisticsCollection.parseDeltaStatsColumnNames(v).isDefined),
+    "needs to be a (possibly empty) comma-separated list of column names.")
```
[the remark this answers is not in the document]
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala:508]: The nested column is represented by `.` notation.
Following are the examples of the nested columns.
```
CREATE TABLE T1 (c0 long, c1 STRUCT <c11: String, c12: long, c13 STRUCT <c131: long, c132: long>>) using delta
TBLPROPERTIES('delta.dataSkippingStatsColumns' = 'c0, c1.c11, c1.c12, c1.c13.c131, c1.c13.c132');

CREATE TABLE T2 (c0 long, c1 STRUCT <c11: String, c12: long, c13 STRUCT <c131: long, c132: long>>) using delta
TBLPROPERTIES('delta.dataSkippingStatsColumns' = 'c0, c1.c11, c1.c12, c1.c13');

CREATE TABLE T3 (c0 long, c1 STRUCT <c11: String, c12: long, c13 STRUCT <c131: long, c132: long>>) using delta
TBLPROPERTIES('delta.dataSkippingStatsColumns' = 'c0, c1');
```

[the remark this answers is not in the document]
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala:508]: Examples of `delta.dataSkippingStatsColumns` are added to description.

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala [hunk 7]
```diff
@@ -389,7 +384,306 @@ trait StatisticsCollection extends DeltaLogging {
...
+   * Note: This method is matching the logical name of tables with the columns inside
+   * DeltaConfigs.DATA_SKIPPING_STATS_COLUMNS. The output of the filter schema is translated into
+   * physical name.
+   *
+   * @param schemaNames the full name path of all columns inside `schema`.
+   * @param schema the original data schema.
+   * @param statsColPaths the specific set of columns to collect stats on.
+   * @param mappingMode the column mapping mode of this statistics collection.
+   * @param parentPath the parent column path of `schema`.
+   * @return filtered schema
```
[the remark this answers is not in the document]
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:593]: removed.

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala [hunk 8]
```diff
@@ -389,7 +384,306 @@ trait StatisticsCollection extends DeltaLogging {
...
+   */
+  def filterSchema(
+      schemaNames: Seq[String],
+      schema: StructType,
+      statsColPaths: Seq[Seq[String]],
+      mappingMode: DeltaColumnMappingMode,
+      parentPath: Seq[String] = Seq.empty): Option[StructType] = {
+    // Find the unique column names at this nesting depth, each with its path remainders (if any)
+    val cols = statsColPaths.groupBy(_.head).mapValues(_.map(_.tail))
```
[the remark this answers is not in the document]
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:602]: Done

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala (hunk 1 above)
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:607]: yes, this is used to handle nested columns.

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala (hunk 2 above)
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:604]: if the field name in the property is `x.y.z`, then `z` is the tail field.
`x.y` is the path to `z`.

This is an example 
```
CREATE TABLE T1 (c0 long, c1 STRUCT <c11: String, c12: long, c13 STRUCT <c131: long, c132: long>>) using delta
TBLPROPERTIES('delta.dataSkippingStatsColumns' = 'c0, c1.c11, c1.c12, c1.c13.c131, c1.c13.c132');
```

File: core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala [hunk 9]
```diff
@@ -389,7 +384,306 @@ trait StatisticsCollection extends DeltaLogging {
...
+   * physical name.
+   *
+   * @param schemaNames the full name path of all columns inside `schema`.
+   * @param schema the original data schema.
+   * @param statsColPaths the specific set of columns to collect stats on.
+   * @param mappingMode the column mapping mode of this statistics collection.
+   * @param parentPath the parent column path of `schema`.
+   * @return filtered schema
+   */
+  def filterSchema(
```
[the remark this answers is not in the document]
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala:595]: Done

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala:508]: few test cases are also added to the code.

**Comment by kamcheungting-db** [comment]
> Would Delta collect stats on the first 32 columns and columns specified in deltaStatsColumns?

These two values are mutual exclusive.
User can only specify either first N columns or named column list.

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala [hunk 10]
```diff
@@ -485,6 +485,23 @@ trait DeltaConfigsBase extends DeltaLogging {
     a => a >= -1,
     "needs to be larger than or equal to -1.")
 
+  /**
+   * The names of specific columns to collect stats on for data skipping. If present, it takes
+   * precedences over dataSkippingNumIndexedCols config, and the system will only collect stats for
+   * columns that exactly match those specified. If a nested column is specified, the system will
+   * collect stats for all leaf fields of that column. If a non-existing column is specified, it
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala:492]:
```suggestion
   * collect stats for all leaf fields of that column. If a non-existent column is specified, it
```

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (kamcheungting-db) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala]: Do you mean I should update the helpMessage of this new table property?
Updated now.

**Review by tdas — APPROVED** [review]
looks great! this is a very highly requested feature in the community, so thank you for doing this.

**Pull request closed without merging**